### ISSUE TITLE
Proposal: Robust dependency management

### ISSUE THREAD

**Issue by hsyed-dojo** [bazelbuild/rules_dotnet#261]
# Robust dependency management

Participants: @hsyed-dojo and @tomdegoede
Reviewers: @purkhusid 

Dependency management is arguably the most complex aspect of managing a Bazel build at scale. A dependency management solution is needed that will work ergonomically in a dual build setup (msbuild and bazel). This is one of 
the ingredients for full build generation #258 and is one of the changes we would like to bring in for increased adoption 
of these official rules #260.

We propose a solution that follows the developer experience, mechanics and patterns in use by 
[rules_jvm_external](https://github.com/bazelbuild/rules_jvm_external).

Dotnet builds have a [convention in the works for centralising the dependencies of a project](https://github.com/NuGet/Home/wiki/Centrally-managing-NuGet-package-versions). A top level
`Directory.Packages.props` is a manifest for all the coordinates and versions in use by a 
solution. We propose that this file become the standard way of encoding dependencies for projects 
that use these rules.

## Core points:

* We propose adding rules and a new Dotnet tool to the project to support this.
* Dependencies will be populated via a repository rule into an external namespace --e.g., `@nuget`.
* Nuget resolution will only be done while preparing/maintaining a lock file. Production 
builds should always be pinned.
* Like `rules_jvm_external` we will have a custom/expanded lock file format to populate 
the dependencies. An expanded format would be preferable because we can then share this with msbuild.
* We aim to use the existing rules and providers to mount the dependencies.

**Comment by samhowes** [comment]
I've done some thinking on this topic over in @rules_msbuild, thought I'd share what I've learned: 

Bazel brings a unique challenge to the table for using Directory.Packages.props:
NuGet packages have different dependency graphs for each target framework that the NuGet package supports. 

Take [NewtonSoft.Json](https://www.nuget.org/packages/Newtonsoft.Json/) for example:
NETStandard 2.0 has no dependencies. 

.NETStandard 1.0 depends on:
  * Microsoft.CSharp (>= 4.3.0)
  * NETStandard.Library (>= 1.6.1)
  * System.ComponentModel.TypeConverter (>= 4.3.0)
  * System.Runtime.Serialization.Primitives (>= 4.3.0)

Directory.Packages.props only specifies the PackageId, but does not specify the frameworks to restore for. If a workspace only uses Netstandard2.0, and an implementation fetches the packages ahead of time without knowing what projects it is fetching for, then a naive implementation could end up fetching a bunch of NuGet dependencies for Netstandard1.0 that arent needed. 

Over in @rules_msbuild, I accommodated this with a similar approach to Directory.Packages.props, but with a slight tweak: I specify the PackageId, and then the list of frameworks that depend on it. This way, packages can be fetched ahead of time without loading build files, and no unnecessary packages are fetched. 

[Example format](https://github.com/samhowes/rules_msbuild/blob/master/deps/nuget.bzl) automatically managed by  [@rules_msbuild//gazelle/dotnet](https://github.com/samhowes/rules_msbuild/tree/master/gazelle/dotnet)
```python
nuget_fetch(
    name = "nuget",
    packages = {
        "NewtonSoft.Json/13.0.1": ["net5.0", "netstandard2.1"], # no transitive dependencies will be fetched
    }
)
```

Given that a Directory.Packages.props file is just another project file to msbuild, this likely could also be handled by simply added metadata to the Package item, so long as that metadata name doesn't conflict with MSBuild's normal metadata for Packages:
```xml
<Project>
    <ItemGroup>
        <PackageReference Include="NewtonSoft.Json" Version="13.0.1">
            <TargetFrameworks>net5.0,netstandard2.1</TargetFrameworks>
        </PackageReference>
    </ItemGroup>
</Project>
```

It would be a slight departure from the standard feature set of the Directory.Packages.props file, but definitely something that an automated tool could manage. [@rules_msbuild//gazelle/dotnet](https://github.com/samhowes/rules_msbuild/tree/master/gazelle/dotnet) would certainly be capable of outputting into this format with a little work on merging the XML files. 

Automation is ideal for this type of Package management that specifies the TargetFramework, because the user would be required to maintain the TargetFramework in two places: once in the project file of the target assembly, and another place for bazel to know what dependency graph to fetch ahead of time. 


Alternatively, the nuget tool could somehow parse all of the project files in the workspace that have packages specified, but that seems to be counter to Bazels approach of Loading, Analysis, and Execution separations. You would also most likely have to actually evaluate the project files, which is starting to encroach on performing an actual build. 

Just my two cents!

**Comment by njlr** [comment]
@purkhusid Would you be able to open-source this tool? I am interested in trying it.

**Comment by purkhusid** [comment]
@njlr Do you specifically want to use Paket or do you just need a better tool for NuGet packages in Bazel? The tool is not currently in shape for sharing it but I might have time soon to clean it up and share it.

We are looking into using a fairly new feature in NuGet which allows centrally managing packages and generating a lock file. If we'll be successful we'll probably make that the default way for 3rd party dependencies in rules_dotnet

**Comment by njlr** [comment]
I am interested in using Paket; I also tried to create a similar tool. Unfortunately my attempts with F# and Bazel failed due to a few other issues (https://github.com/bazelbuild/rules_dotnet/issues/234, https://github.com/bazelbuild/rules_dotnet/issues/235)

**Comment by purkhusid** [comment]
@njlr Ok, I'll take a look at this as soon as I can. My bandwidth is fairly limited at the moment but I hope I can do some rules_dotnet work soon.

**Comment by purkhusid** [comment]
The Paket support works very well at the moment and is now on `master`. Me and @tomdegoede have been looking into adding support for parsing a central `Directory.Packages.props`(https://devblogs.microsoft.com/nuget/introducing-central-package-management/) but it's still WIP.

**Comment by Place1** [comment]
I've been looking at implementing a solution based on `packages.lock.json` [microsoft blog post](https://devblogs.microsoft.com/nuget/enable-repeatable-package-restores-using-a-lock-file/).

I was able to write a small script that converts the lock file into a `nuget_repo` but I found I had to manually update the sha512 value for each dependency because of [this implementation detail](https://github.com/NuGet/Home/issues/10787) regarding the `contentHash` calculation from nuget.

Just wanted to post this in case someone's interested in this approach as an alternative to paket.
